File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummyNode = new ListNode();
        ListNode tail = dummyNode;

        while(list1 != null && list2 != null) {

            //append list1's head to the list by pointing tail's next pointer 
            //to list1's head
            if(list1.val < list2.val) {
                tail.next = list1;
                list1 = list1.next;
            }

            //do the same if list2 is smaller than list1
            else {
                tail.next = list2;
                list2 = list2.next;
            }

            //move tail pointer forward one node
            tail = tail.next;
        }

        //append the rest of either lists if nodes left
        if(list1 != null) {
            tail.next = list1;
        }
        else {
            tail.next = list2;
        }

        //returns the head of the resulting list
        return dummyNode.next;
    }
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """

        # make a dummy node that equals a tail pointer and set the dummy node value to 0
        dummyNode = tail = ListNode()

        # while list1 and list2 is not empty
        while list1 and list2:

            # compare list1 and list2 value and take smaller node
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next

            else:
                tail.next = list2
                list2 = list2.next

            # set the tail's pointer to the smaller node
            tail = tail.next

        # if list1 or list2 still has values, point to tail
        if list1:
            tail.next = list1
        else:
            tail.next = list2

        return dummyNode.next        